packages/noya-state/src/reducers/styleReducer.ts

```diff
@@ -280,6 +308,97 @@ export function styleReducer(
         draft.borders[index].position = position;
       });
     }
+    case 'setFillFillType':
+    case 'setBorderFillType': {
+      const [, index, type] = action;
+
+      const category = action[0] === 'setFillFillType' ? 'fills' : 'borders';
+
+      return produce(state, (draft) => {
+        if (!draft[category] || !draft[category]![index]) return;
+        draft[category]![index].fillType = type;
+
+        if (!draft[category]![index].gradient) {
+          draft[category]![index].gradient = Models.fill.gradient;
+        }
+      });
+    }
+    case 'setFillGradientColor':
+    case 'setBorderGradientColor': {
+      const [, index, stopIndex, color] = action;
+      const category =
+        action[0] === 'setFillGradientColor' ? 'fills' : 'borders';
+
+      return produce(state, (draft) => {
+        if (
+          !draft[category] ||
+          !draft[category]![index] ||
+          !draft[category]![index].gradient ||
+          !draft[category]![index].gradient.stops[stopIndex]
+        )
+          return;
+
+        draft[category]![index].gradient.stops[stopIndex].color = color;
+      });
+    }
+    case 'setFillGradientPosition':
+    case 'setBorderGradientPosition': {
+      const [, index, stopIndex, position] = action;
+      const category =
+        action[0] === 'setFillGradientPosition' ? 'fills' : 'borders';
+
+      return produce(state, (draft) => {
+        if (
+          !draft[category] ||
+          !draft[category]![index] ||
```

Comment:
We typically want to avoid using `!` to ignore null/undefined values, since it's much easier to cause a crash. If we can't get the types to work out, we should make 2 separate `case`s.

Separately, maybe we could make a `gradientReducer`? And then pass either fill or border gradient into it? I think that'll result in shorter code still.

Reply:
Yeah! I was thinking that too